pkg/providers/vsphere/internal/templates/factory.go

```diff
@@ -148,3 +167,143 @@ func (f *Factory) importOVAIfMissing(ctx context.Context, templateName, ovaURL s
+				alreadySized = true
+			}
+		} else if strings.EqualFold(deviceLabel, "Hard disk 2") {
+			disk2 = deviceInfo.(map[string]interface{})["Name"].(string)
+			parsedDiskSize := parseDiskSize(deviceDiskSize)
+
+			if parsedDiskSize == -1 {
+				logger.V(2).Info("Unable to get converted KB to GB int")
+				return fmt.Errorf("unable to get converted KB to GB int: %d", parsedDiskSize)
+			}
+
+			if parsedDiskSize == 20 {
+				alreadySized = true
+			}
+			break
+		}
+	}
+
+	if alreadySized {
+		return nil
+	}
+	diskName := ""
+	var diskSizeInGB int
+	if disk2 != "" {
+		logger.V(2).Info("Resizing disk 2 of template to 20G")
+		diskName = disk2
+		diskSizeInGB = 20
+	} else if disk1 != "" {
+		logger.V(2).Info("Resizing disk 1 of template to 22G")
+		diskName = disk1
+		diskSizeInGB = 22
+	} else {
+		return fmt.Errorf("template %v is not valid as there are no associated disks", templateName)
+	}
+
+	//logger.V(2).Info("VM Info. ", "vm info", response.String())
+
+	logger.V(2).Info("Template Name. ", "Template name: ", templateName)
+	logger.V(2).Info("Printing diskSize for resize: ", "Disk Size: ", diskSizeInGB)
+	err = f.client.ResizeDisk(ctx, datacenter, templateName, diskName, diskSizeInGB)
+	if err != nil {
+		return fmt.Errorf("resizing disk %v to 20G: %v", diskName, err)
+	}
+
+	devicesInfo2, err := f.client.DevicesInfo(ctx, datacenter, templateName)
+	if err != nil {
+		return err
+	}
+
+	for _, deviceInfo := range devicesInfo2.([]interface{}) {
+		deviceMetadata := deviceInfo.(map[string]interface{})["DeviceInfo"]
+		deviceDiskSize := deviceMetadata.(map[string]interface{})["Summary"].(string)
+		deviceLabel := deviceMetadata.(map[string]interface{})["Label"].(string)
+
+		if strings.EqualFold(deviceLabel, "Hard disk 1") || strings.EqualFold(deviceLabel, "Hard disk 2") {
+			parsedDiskSize := parseDiskSize(deviceDiskSize)
+
+			if parsedDiskSize == -1 {
+				logger.V(2).Info("Unable to get converted KB to GB int")
+				return fmt.Errorf("unable to get converted KB to GB int: %d", parsedDiskSize)
+			}
+
+			if disk2 != "" {
+				//check if disk1 is 20gb
+				if parsedDiskSize != 20 {
+					return fmt.Errorf("error resizing disk1 to 20 GB - new size is: %d", parsedDiskSize)
+				}
+			} else if disk1 != "" {
+				//check if disk2 is 22 gb
+				if parsedDiskSize != 23 {
+					return fmt.Errorf("error resizing disk2 to 22 GB - new size is: %d", parsedDiskSize)
+				}
+			} else {
+				return fmt.Errorf("template %v is not valid as there are no associated disks", templateName)
+			}
+		}
+
+	}
+
+	//templateFullPath := filepath.Join(templateDir, templateName)
+
+	logger.V(4).Info("Taking template snapshot", "templateName", templateFullPath)
+	if err := f.client.CreateVMSnapshot(ctx, datacenter, templateFullPath); err != nil {
+		return err
+	}
+
+	logger.V(4).Info("Marking vm as template", "templateName", templateFullPath)
+	if err := f.client.MarkVMAsTemplate(ctx, datacenter, templateFullPath); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
I would suggest revisiting this whole method. It's quite lengthy and has a lot of nested branches, it could use a small refactor